### PR TITLE
Prevent rare foreign key error "log_broadcasts_eth_log_id_fkey"

### PR DESCRIPTION
**Pull request by samsondav**
In some very rare cases if we delete an eth_log whilst simultaneously
upserting a log broadcast we can insert stale data (an eth_log id that
no longer exists) causing foreign key errors.

In this case we use optimistic locking such that if we receive this
foreign key violation we act as if the log was inserted successfully
then immediately deleted again.

### CODE REVIEW

File: core/services/log/orm.go [hunk 1]
```diff
@@ -152,7 +163,18 @@ ON CONFLICT (%[1]s, block_hash, log_index) WHERE %[1]s IS NOT NULL DO UPDATE SET
 	}
 
 	stmt := fmt.Sprintf(q, jobIDName)
-	return o.db.Exec(stmt, args...).Error
+	err = o.db.Exec(stmt, args...).Error
+	switch v := err.(type) {
+	case *pgconn.PgError:
+		if v.Code == "23503" && v.ConstraintName == "log_broadcasts_eth_log_id_fkey" {
+			// NOTE: This rare case indicates that the eth_log was deleted
```
Comment (j16r) [inline core/services/log/orm.go:170]: Should we log here? (and above?) Maybe a warning?

Comment (samsondav) [reply core/services/log/orm.go:170]: I'm not really sure it indicates any type of error, I think it's expected occasionally in normal operation.

Comment (samsondav) [reply core/services/log/orm.go:170]: @spooktheducks can you confirm?

Comment (samsondav) [reply core/services/log/orm.go:179]: I spoke with spook, this can legitimately happen but I pushed debug logging anyway

**Review by j16r — APPROVED** [review]

**Pull request merged**